Fix G33 + report and calc delta calibration radius

LVD-AC opened the pull request:
Add `dcr` to the screen output of G33 for user reference.

Little add-on to #22707

New more precise calculation of the default calibration radius for offset probes at nozzle positions

![radius](https://user-images.githubusercontent.com/26793057/133923992-c661963d-1576-4cd8-8853-faf374850bac.png)

LVD-AC committed:
Report delta calibration radius + calc calibration radius

Add dcr to the screen output of G33 for user reference.

Calc dcr

Calculate calibration radius for offset probes by using new config parameter BED_RADIUS

oops

oops

oops

Update G33.cpp

Update G33.cpp

LVD-AC commented on Marlin/src/gcode/calibrate/G33.cpp:
```diff
@@ -409,17 +409,23 @@ void GcodeSuite::G33() {
                   towers_set = !parser.seen_test('T');
 
-  float max_dcr = dcr = DELTA_PRINTABLE_RADIUS;
+  // needs to be moved to all delta config files
```
needs work, do not merge yet

LVD-AC commented:
![radius](https://user-images.githubusercontent.com/26793057/133880242-5748330d-5b74-4f34-a4b6-948b35c7dcc6.png)

- set Calibration radius to Printable radius
- If the Probe offset < (Bed radius/2 - Printable radius) leave as it is
- Else reduce by Excess = Probe offset  - (Bed radius/2 - Printable radius)

LVD-AC commented on Marlin/src/gcode/calibrate/G33.cpp:
```diff
@@ -409,17 +409,21 @@ void GcodeSuite::G33() {
                   towers_set = !parser.seen_test('T');
 
+  // needs to be moved to all delta config files
+  #define BED_DIAMETER 220
```
Is there a way to do this in one go?

thinkyhead replied:
The bed's diameter is assumed to be the `X_MAX_POS - X_MIN_POS` (`X_MAX_LENGTH`) which should be equal to the `Y_MAX_POS - Y_MIN_POS` (`Y_MAX_LENGTH`) on a delta. Then the `DELTA_PRINTABLE_RADIUS` (and apparently `X_BED_SIZE`, `Y_BED_SIZE`) are smaller than this diameter.

LVD-AC replied:
On deltas : 
(Configuration.h)
```
#define X_BED_SIZE ((DELTA_PRINTABLE_RADIUS) * 2)
#define Y_BED_SIZE ((DELTA_PRINTABLE_RADIUS) * 2)

#define X_MIN_POS -(DELTA_PRINTABLE_RADIUS)
#define Y_MIN_POS -(DELTA_PRINTABLE_RADIUS)
#define X_MAX_POS DELTA_PRINTABLE_RADIUS
#define Y_MAX_POS DELTA_PRINTABLE_RADIUS
```
and
(Conditional_post.h)
```
#define X_MAX_LENGTH (X_MAX_POS - (X_MIN_POS))
#define Y_MAX_LENGTH (Y_MAX_POS - (Y_MIN_POS))
```
so `X_MAX_LENGTH` and `Y_MAX_LENGTH` are also set to `((DELTA_PRINTABLE_RADIUS) * 2)` and as such are always equal by definition to `X_BED_SIZE` and `Y_BED_SIZE`.

I need to know the size border on the bed that can't be used for printing but that can be used for (offsetted) probing at nozzle positions (blue area).

<details><summary>The Image Again</summary>

![radius](https://user-images.githubusercontent.com/26793057/133888540-6cd76454-7f97-4029-99b6-54b0316a2562.png)
</details>

thinkyhead replied:
The old setting `DELTA_PROBEABLE_RADIUS` was replaced with the `DELTA_PRINTABLE_RADIUS - PROBING_MARGIN`, indicating the "region of the bed" which is able to be probed (because getting too close to the edge with an inductive or capacitive probe will produce bad readings).

Meanwhile, the functions `position_is_reachable` (`motion.h`) and `probe.can_reach` (`probe.h`) are used to check the physical limits on Delta motion. Since the probe offsets and the software endstops can both be changed at runtime, these functions are used instead of relying on some fixed value. The `position_is_reachable` function takes an extra `inset` parameter.

The reachable area outside of the bed on a delta might end up being required as a new setting, but it would be good to first consider whether the same thing can be accomplished with existing settings. Specifically, the `*_MIN_POS` / `*_MAX_POS` values should have a corresponding delta setting (e.g., by adding `BED_DIAMETER`) and the `*_BED_SIZE` values should have a corresponding delta setting (currently, `DELTA_PRINTABLE_RADIUS`). Since the existing setting is a radius, probably the new one should also be expressed as a radius (e.g., "`MAX_RADIUS`" instead of "`BED_DIAMETER`").

And, it will be important to give some thought to `position_is_reachable` and `probe.can_reach` with any updates to delta settings.

thinkyhead replied:
> Is there a way to do this in one go?

With some text editors (e.g., Sublime) you can do a global search-and-replace only within the `examples/delta` folder, searching for the line above (or below) the new addition, and replacing that line with: the existing line plus the new addition. However, the new line will not have the same value in all cases, so you would follow it up by editing all the added lines to give them the proper values.

e.g., Replace…
```
(( +)#define )DELTA_PRINTABLE_RADIUS( +)([^ ]+)( *)(.*)
```
With…
```
$1DELTA_PRINTABLE_RADIUS$3$4$5$6
$2// Maximum reachable area
$1DELTA_MAX_RADIUS      $3$4$5// (mm)
```

thinkyhead committed:
Use "MAX_RADIUS" and fix indent

thinkyhead committed:
clarify a difference

thinkyhead commented on Marlin/src/gcode/calibrate/G33.cpp:
```diff
@@ -409,17 +409,21 @@ void GcodeSuite::G33() {
+    else {
+      // else with nozzle positions only the nozzle needs to be within the printable area
+      // and the probe needs to be inside the physical machine diameter
+      max_dcr = dcr -= _MAX(HYPOT(probe.offset_xy.x, probe.offset_xy.y) - (MAX_RADIUS) + (DELTA_PRINTABLE_RADIUS), 0);
```
A probe which is, for example, offset to the left of the nozzle can always reach the left edge, but maybe not the right edge. The `probe.can_reach` method takes this into consideration instead of applying the limit to the entire perimeter as a reduction in radius. Is it important in `G33` for the probing to be symmetrically centered around 0,0 ?

LVD-AC replied:
1) "Since the probe offsets and the software endstops can both be changed at runtime"

I got that (G33 changes the endstops itself). So having an EEPROM setting for calibration radius is no longer a good idea; That's why I try to calculate it at the start of G33 taking into account the current probe offsets.

2) "The reachable area outside of the bed on a delta might end up being required as a new setting"

I think so as well. Delta calibration is the only occurence were probing on nozzle positions is required, and it's only then that extra margin outside the printable area can be used (and needs to be known).

3) "*_MIN_POS / *_MAX_POS values should have a corresponding delta setting"

*_MIN_POS / *_MAX_POS are quit intensely used by G29 and other stuff and currently are linked to DELTA_PRINTABLE_RADIUS. IMO in order not to disturb the current use of current parameters, it would be better not to link the new parameter to anything since as stated above it's only used for nozzle position probing.

3) "MAX_RADIUS" instead of "BED_DIAMETER"

OK with that. Maybe it would be even better not to refer to this as a radius or diameter but as a margin like EXTRA_MARGIN_FOR_PROBING, since it only appies to the side the probe is mounted and not to the whole bed.

5) "And, it will be important to give some thought to position_is_reachable and probe.can_reach"

The old testing in G33 was done by testing each point with position_is_reachable; but since it is a circular grid you might as well test dcr against one calculated number.

6) "Is it important in G33 for the probing to be symmetrically centered around 0,0"

For the moment YES. For the match to work (especially with the added probe positions option) it requires symmetry on all 3 axis.

thinkyhead committed:
etc.

thinkyhead committed:
DELTA_MAX_RADIUS to examples

thinkyhead committed:
Delta compatibility

thinkyhead committed:
Tweak max_dcr update

thinkyhead commented:
`DELTA_MAX_RADIUS` is now added to the delta configs. That should be useful to allow a wider probing radius for offset probes too. I'll have to play around with that. I'll have to see if I can charm AnyCubic into sending me a Kossel or a Predator.

LVD-AC committed:
Only delta calibration

conditionals

Update Conditionals_post.h

oops

oops

Update Conditionals_post.h

oops

Probe testing

Probe should be allowed outside printable area - as long as the nozzle is within printable area we're fine

Revert "Probe testing"

This reverts commit 5293c811e028dd93b3a2384a02faa8946db701c6.

Update Conditionals_post.h

LVD-AC commented:
"DELTA_MAX_RADIUS is now added to the delta configs"

Thanks for that.

`// Maximum reachable area` maybe add to it **" = farthest point outward in the direction of the offset where an offset-probe runs off the bed and no longer triggers correctly"**

Shouldn't DELTA_MAX_RADIUS be commented out for printers with central probes?

LVD-AC committed:
reorder calc

reoder2

oops

calc bis

Update G33.cpp

Update G33.cpp

Update G33.cpp

LVD-AC commented:
`DELTA_MAX_RADIUS` does **not** indicate the area where **the effector** can move, but where **the probe** can be, while the effector stays withing the printable area. Effector (=nozzle) moves (printing and non-printing) are limited by definition to `DELTA_PRINTABLE_RADIUS`.

Anyhow G33 does not take into account `DELTA_MAX_RADIUS` on printers that have `HAS_PROBE_XY_OFFSET` disabled. `DELTA_MAX_RADIUS` is a new parameter that is linked to **probing at nozzle positions** and only G33 does that.

LVD-AC commented:
May I suggest the following for the config files :

```
  #if ENABLED(DELTA_AUTO_CALIBRATION)
  // Maximum reachable area - = farthest point outward where the probe runs off the bed or no longer triggers correctly
  // Uncomment if an offset probe is present and set at least to DELTA_PRINTABLE_RADIUS
  //#define DELTA_MAX_RADIUS       85.0     // (mm)
  #endif

```
And if someone finds another use for it (I doubt that will happen), the config files, sanety check and conditional post will need to be changed accordingly. But as such, it is clear it is a parameter that (as it stands) is used only for delta calibration with offset probes.

LVD-AC commented:
I did run some tests on my printer by forcing the central probe to probe with an offset:

central grid:

calibration radius = 80   number of iterations 5 std dev 0.010
calibration radius = 60   number of iterations 4 std dev 0.008
calibration radius = 40   number of iterations 4 std dev 0.006

As long as the grid is centered correcty the routine comes to decent calibration in 4-5 iterations even when the calibration radius is less than half of what is possible with a central probe (85mm).

offsetted grid:

calibration radius = 40  X-offset = 40 number of iterations 9 std dev 0.013 (O option with probe offset 40mm)
calibration radius = 72.5  X-offset = 12.5 number of iterations 9 std dev 0.09 (the max option above with probe offset 40mm)

The routine realy struggles to get rid off those false tilts and to get a good calibration even with a small offset, but it gets there.

IMO the Max option and even the O option are useless : less accurate calibration due to the false tilts and it takes twice as long.

@thinkyhead it's your call if we leave the O-option in or not for those who insist in doing things wrong (and for us to avoid having to answer issues on that over and over), but implementing another M(ax)-option (probing somewhere in between nozzle and probe) will only confuse since it doesn't add anything usefull. That was a bad idea!!!

LVD-AC commented on Marlin/src/module/probe.cpp:
```diff
@@ -767,13 +767,9 @@ float Probe::probe_at_point(const_float_t rx, const_float_t ry, const ProbePtRai
   // On delta keep Z below clip height or do_blocking_move_to will abort
   xyz_pos_t npos = { rx, ry, TERN(DELTA, _MIN(delta_clip_start_height, current_position.z), current_position.z) };
   if (probe_relative) {                                     // The given position is in terms of the probe
-    if (!can_reach(npos)) {
```
We need to disable this, now that we allow probing outside the printable area

thinkyhead committed:
indent

thinkyhead commented:
> it's your call if we leave the O-option in or not for those who insist in doing things wrong

If it's only causing confusion then I think it would be fine to remove it. Is there some situation where you anticipate someone will really miss it?

thinkyhead commented:
> `DELTA_MAX_RADIUS` does not indicate the area where the effector can move (limited by the effector running into the towers), but where the probe can be (limited by the bed size), while the effector stays [within] the printable area. 

I think that it should take on that role. It seems strange to be using a fixed value for the max radius and to apply it exclusively in `M33` when the nozzle-to-probe offset can be changed at runtime. Wouldn't it make more sense to consider the `DELTA_MAX_RADIUS` the maximum area that the effector can cover, and then apply a constraint based on the current nozzle-to-probe offset? We would also need to know how far the probe (and its mount) can be off-center on the effector before it has any effect on the width of the effector and therefore requires more constraint.

With `DELTA_MAX_RADIUS` acting as the limit on the total motion of the effector, it is easy to determine whether the probe can reach a certain point, because we can check that the effector would end up beyond the `DELTA_MAX_RADIUS` if the probe were moved to that point.

In my opinion, a new and valuable `DELTA_MAX_RADIUS` setting should not just be used in `M33` but should also be used in `position_is_reachable` as a more accurate gauge for whether the nozzle / effector can move to a certain point anywhere inside or outside the bed. The "printable area" radius is then only used to choose the default min/max software endstops.

If `DELTA_MAX_RADIUS` is not used in the manner I propose, is there some other method to check the total movement range of the effector, such as calculating the "rounded triangle" that it can reach based on length of the arms, etc.?

LVD-AC commented:
1) "Is there some situation where you anticipate..." there allways have been quit a lot of issues demanding calibration should be done at probe positions like in G29 and everytime we need to explain the difference between calibration and bed leveling : bed leveling assumes the kinematic moves are in a plane and searches for bed imperfections so probing needs to be done at **bed positions**/ calibration assumes the bed is flat and level and adjusts the kinematic moves of the effector to be parallel to it so probing needs to be done at **kinematic positions** of the effector. The only reason to keep the O-option in is for those who insist to calibrate by probing bed positions: it's wrong and it will be less accurate (but by some chear luck it works as well) so be my guest. I've left the option in just to make everybody happy.

2) In bed levelling it makes no sense to search and correct for bed imperfections outside the print area. But if someone can find some other use for `DELTA_MAX_RADIUS` be my guest, just make sure it does not compromise the G33 use it was intended for : define the limit **the probe can move outside 'DELTA_PRINTABLE_RADIUS'**. So it is basicly half the bed diameter minus mounting holes, clamps and a margin for the probe to work properly.

If you're gonna move the effector upto `DELTA_MAX_RADIUS` it will surely run into the towers and possibly do some damage. That is why the calculated dcr in G33 is topped with 'NOMORE(dcr, DELTA_PRINTABLE_RADIUS)' at the end, to make sure **the effector stays within `DELTA_PRINTABLE_RADIUS` at all times**.

LVD-AC commented:
The best reason to keep the G33 O option in : These offsets are a recurrent issue; so it is only a matter of time someone puts in a PR to change them again. But if offsets are offered as an option there is no need to delete the right way of doing things.

The best reason to remove the G33 O option : It educates people that doing things wrong is OK as well. Why else would it be implemented? (answer: see the best reason to keep it in) ;)

LVD-AC commented:
"is there some other method to check the total movement range of the effector, such as calculating the "rounded triangle" that it can reach based on length of the arms, etc"

This has nothing to do with the effector or the arms, but only with the size of the printbed. The movement range of the effector stays unaltered and is limited to `DELTA_PRINTABLE_RADIUS` at all times.

"Usable bedsize" is the new parameter. IMO the least confusing way to explain to users what this is : 

in configuration.h

```
// set the diameter of the printbed (as printbeds are always expressed in diameter)
#define BED_DIAMETER 220 //mm
// set the margin on the printbed for the probe to work propperly (mounting holes, clamps...) 
#define BED_MOUNTING_MARGIN 10 //mm
```

and calculate the necessary from there.

note:

Anyhow G33 Rx.xx would be best redifined to temporay **R**educe the radius of the calibration grid by xx.x mm as `dcr` (delta calibration radius) is now an internal variable of wich the lamdba user who is not familiar with older Marlin versions has no knowledge of, knows the definition of it or the way it is calculated.

LVD-AC committed:
Redef R option

LVD-AC committed:
Merge branch 'G33-report' of https://github.com/LVD-AC/Marlin into G33-report

LVD-AC committed:
Revert "Merge branch 'G33-report' of https://github.com/LVD-AC/Marlin into G33-report"

This reverts commit a0c51f9c6035b0b554a3bad68edc5bc24f5eb4b3, reversing
changes made to cd4851059e86f0878cf36b10a8db1f2322ecef8c.

LVD-AC committed:
Revert "Revert "Merge branch 'G33-report' of https://github.com/LVD-AC/Marlin into G33-report""

This reverts commit eb3a52a2b8a5766f1b16a16c678503ac5f81fe7e.

LVD-AC committed:
Redef R-option

Redef R-option

R-option

Use bed_diameter

reorder

oops

LVD-AC commented:
@thinkyhead if you do not agree to use `BED_DIAMETER` as parameter in the config files just revert the last commit. If we keep DELTA_MAX_RADIUS we'll just have to state in the documentation that this half the bed diameter minus a margin for mounting holes or clamps and has nothing to do with the limitions on the effector moves.

LVD-AC committed:
use BED_DIAMETER

Foxies-CSTL commented:
Hello everyone,
After merging the PR #22707, I can't do Delta calibration anymore. The movement stops at three circular polls in the center and it no longer goes to the edge of the bed. So before posting the problem, I did some research on the current PR and looked at the difference between the files. By importing the three old files (G33.cpp, delta.h, delta.cpp) into the latest BugFix, I found the normal calibration sequence. Anyway, I read this last PR and I integrated the different modifications on the concerned files, I just made a compilation and a test on my printer. The result is that the calibration sequence (G33 or G33 P4) stops with a reset (HOMING Failed) after several probings in the center of the bed.

https://user-images.githubusercontent.com/24864147/134541308-213c400b-aa3e-4371-937a-733134952941.mp4

LVD-AC commented:
It looks like the calculated calibration radius is set to zero for some reason or another. Have you tried if G33 Rx.xx works?

Foxies-CSTL commented:
> It looks like the calculated calibration radius is set to zero for some reason or another. Have you tried if G33 Rx.xx works?

Hi @LVD-AC 
No, I will try (G33 R130.00) later (printing in progress) but it works on your Delta with the latest BugFix?

thinkyhead commented:
The addition of `BED_MOUNTING_MARGIN` is an interesting choice. Presumably the width of the effector, the length of the arms, and the distances between the towers can all be combined to calculate the maximum reachable area in terms of the nozzle position. But I suppose it is possible that there could be other obstructions that need to be considered.

LVD-AC committed:
solve bug report

LVD-AC commented:
@Foxies-CSTL I think I found the culprit: double definition of dcr; the last commit here should take care of that.

LVD-AC committed:
can_reach

include absolute/relative parameter like in probe_at_point

bug

LVD-AC commented:
@thinkyhead `DELTA_PRINTABLE_RADIUS` takes care of all obstructions in all directions what the effector is concerned. DELTA_PRINTABLE_RADIUS is by definition the largest radius the nozzle can reach without any obstruction.

The new parameter `BED_DIAMETER` (+`BED_MOUNTING_MARGIN`) is simply the size of the bed (nomatter if it is reachable or not by the effector). E.g. the printer has a heated bed of 220mm diameter with mounting holes at 10mm or a glass bed of 200mm diameter with clamps of 3mm. Since bed sizes are most commonly expressed in diameter I prefer this over radius. This is IMO the most clear and simple definition that does not require 2 pages of documention to explain what it is.

Any routine (like G33) that wants to make use of the entire bed diameter can do so with `probe_at_point` and `can_reach` by including the `probe_relative` parameter.

Foxies-CSTL commented:
> 
> 
> @Foxies-CSTL I think I found the culprit: double definition of dcr; the last commit here should take care of that.
> 
> Could you check PR #22707 by changing in G33.cpp
> 
> ```
>   float max_dcr = dcr = DELTA_PRINTABLE_RADIUS;`
> ```
> 
> to
> 
> ```
>   float max_dcr;
>   max_dcr = dcr = DELTA_PRINTABLE_RADIUS;
> ```
Hi,
Application of the patch proposed on the last BugFix merge. Same problem in the calibration sequence (G33 or G33 R130.00). The beginning of the sequence is good but stops after 3 probes at 5cm from the center then goes back up for the end of the sequence. No measurement towards the edge of the bed (#define PROBING_MARGIN 15)

[20210923-G33-FixCalib-PR22795.txt](https://github.com/MarlinFirmware/Marlin/files/7231457/20210923-G33-FixCalib-PR22795.txt)

[20210923-G33R130-FixCalib-PR22795.txt](https://github.com/MarlinFirmware/Marlin/files/7231458/20210923-G33R130-FixCalib-PR22795.txt)

Foxies-CSTL commented:
Correction:
I managed to perform a complete sequence (only 1 iteration) with G33 R100.00.

LVD-AC commented:
@Foxies-CSTL and what happens after the 1st iteration?

Foxies-CSTL commented:
> 
> 
> @Foxies-CSTL and what happens after the 1st iteration? Can you do the same with G33 R129.0?

Same result as the G33 command: simple 3 middle probes and end of sequence.
[20210926-G33R129.0-FixCalib-PR22795.txt](https://github.com/MarlinFirmware/Marlin/files/7231620/20210926-G33R129.0-FixCalib-PR22795.txt)

Foxies-CSTL commented:
> Can you do the same with G33 R129.0?
I managed to get 4 full iterations using G33 P4 R80.00 V3 but on the last iteration it only does 3 probes (as with G33) and ends the sequence with a valid result.

LVD-AC commented:
I installed a newer version of Arduino and now 2.0 compiles again, so I can run some tests off my own now.

Foxies-CSTL commented:
> 
> 
> I installed a newer version of Arduino and now 2.0 compiles again, so I can run some tests off my own now.

You can also switch to VSC or VSCodium with PlatformIO, it's easier. ;-)

LVD-AC commented:
@Foxies-CSTL  I found another bug in #22707:
In `G33.cpp` replace

```
dcr /= 0.9f;
```

with
```
if (_7p_9_center) dcr /= 0.9f;
```

Foxies-CSTL commented:
> @Foxies-CSTL I found another bug in #22707:
> In `G33.cpp` replace
> ```
> dcr /= 0.9f;
> ```
> with
> ```
> if (_7p_9_center) dcr /= 0.9f;
> ```

Hi,
Same problem with G33, only three probings and it's over :-(

LVD-AC committed:
PROBE_MARGIN negative

Allow PROBE_MARGIN to be negative.

LVD-AC commented:
@thinkyhead: If we allow `PROBE_MARGIN` to be negative for offsetted probes as well it can be used to indicate that extra bed area that is available for probing and as such we don't need an extra parameter to calculate `dcr` correctly. But I do not know if this would obset other routines that maybe rely on `PROBE_MARGIN` always being positive.

Foxies-CSTL commented:
> 
> 
> @thinkyhead: If we allow `PROBE_MARGIN` to be negative for offsetted probes as well it can be used to indicate that extra bed area that is available for probing and as such we don't need an extra parameter to calculate `dcr` correctly. But I do not know if this would obset other routines that maybe rely on `PROBE_MARGIN` always being positive.

Also pay attention to the margin used because it can be minimal (5) if the stallguard2 (sensorless_probing) function is used with the TMC2209.

LVD-AC commented:
In case of negative `PROBE_MARGIN` the comment in the config files need to be changed:

`// with NOZZLE_AS_PROBE this can be negative for a wider probing area.`
 to
`// this can be negative for a wider probing area.`

LVD-AC commented:
Running tests now; so far so good, the routine runs on maximum radius upto the end:

Central probe + PROBE_MARGIN = 0 ... OK dcr = DELTA_PRINTABLE_RADIUS
Central probe + PROBE_MARGIN > 0 ... OK dcr = DELTA_PRINTABLE_RADIUS - PROBE_MARGIN
Central probe + PROBE_MARGIN < 0 ... OK dcr = DELTA_PRINTABLE_RADIUS

I'll simulate an offset probe now.

Foxies-CSTL commented:
Okay, I applied all the modifications of this PR22795 (5 files) and run a G33 calibration. Success, I get complete iterations with results that are refined as the iterations. 
Results:
`Recv: Calibration OK    rolling back.
Recv: //action:notification Calibration sd:0.163
Recv: .Height:335.07  Ex:-1.33  Ey:-0.63  Ez:+0.00  Radius:134.29
Recv: .               Tx:-0.16  Ty:-0.19  Tz:+0.35`
Thanks for your work done.

LVD-AC commented:
Tests with offset (X=-20, Y=15, total 25):

Offset probe + PROBE_MARGIN = 0 ... OK                    dcr = DELTA_PRINTABLE_RADIUS - offset
Offset probe + PROBE_MARGIN = 0 + O-option ... OK dcr = DELTA_PRINTABLE_RADIUS - offset
Offset probe + PROBE_MARGIN > 0 ... OK                     dcr = DELTA_PRINTABLE_RADIUS - offset - PROBE_MARGIN
Offset probe + PROBE_MARGIN > 0 + O-option ... OK dcr = DELTA_PRINTABLE_RADIUS - offset
Offset probe + PROBE_MARGIN < 0 ... OK                     dcr = DELTA_PRINTABLE_RADIUS - offset - PROBE_MARGIN
Offset probe + PROBE_MARGIN < 0 + O-option ... OK dcr = DELTA_PRINTABLE_RADIUS - offset

All tests are OK.

Foxies-CSTL commented:
Just an observation on "PROBING_MARGIN" and "Multiple Probing= not used": The first iteration runs with a margin of 5 and two successive probing for each point then on the second iteration, the defined variable (15 for me) is used as well as a single probing as defined in my configuration

LVD-AC commented:
If @thinkyhead is OK with this, I'm done.

The routine does what it expected; users who have extra bed space available can use it for probing outside `DELTA_PRINTABLE_RADIUS` by setting `PROBE_MARGIN` negative (only on kinematic printers). The effector will stay in the safe zone without running into the towers. 

Developers that want to use this feature for other purposes can use `probe_at_point()` and can_`reach()` by setting the `probe_relative` parameter to `false`.

No extra parameters required in the config files. (MAX_RADIUS can be deleted again)

LVD-AC commented:
@Foxies-CSTL 
> and two successive probing for each point

 = normal behavior. I've noticed that in your video by counting the N° of probes; this is because your hardware default `DELTA_HEIGHT` in your config file is set to low : the probe descends and runs in the bed before the slow probe move takes place, so the probe retracts and probes again. In the seccond run `DELTA_HEIGHT` is already corrected so this no longer occurs.

> The first iteration runs with a margin of 5 ... on the second iteration, the defined variable (15 for me) is used

??? In all my tests the probing circle did not change from one iteration to another, and the calibration radius was set according to 
`PROBE_MARGIN` as set in configuration.h.

Foxies-CSTL commented:
> ??? In all my tests the probing circle did not change from one iteration to another, and the calibration radius was set according to
> `PROBE_MARGIN` as set in configuration.h.

I confirm my observation on the first iteration: double probing of all points and the margin at the edge of the bed is smaller than defined. The following iterations are consistent with the values defined by the configuration. I redid the test with "G33 P4 V3".

Foxies-CSTL commented:
> 
> 
> I presume your initial settings are quit off:
> 
> ```
> #define DELTA_HEIGHT 380.0
> #define DELTA_RADIUS 140.8
> ```
> 
> Correct your default values in your `configuration.h` file to more or less what you get after calibration, and this odd behavior should dissapear.

Sure! but the parameters that trigger this first special iteration are the height "DELTA_HEIGHT" and the endstop "DELTA_ENDSTOP_ADJ". 

Well, I just applied your last changes (4 files) and run a calibration.
Everything seems ok.
Thanks for your work.

LVD-AC commented:
Thanks for the testing.

BTW: if DELTA_RADIUS is off there will be dimentional errors, so it only seems the routine is using a different PROBE_MARGIN. That is why some people requested to temporary reduce the calibration radius for a 1st run to make sure everything stays within limits and do a final run on the maximum radius possible, That's now avaible with the R(educe) option in G33.

Foxies-CSTL commented:
> Thanks for the testing.

You're welcome, but it shows once again that modifications to Marlin sources, without physical testing on a Delta printer, have malfunctioning consequences despite the good intentions of the developers.

@thinkyhead , try contacting FLSun to get his latest Delta the SR. It is a very good machine. I already contacted them for this but I don't have enough weight despite my support of Marlin on the QQS and Q5.

LVD-AC committed:
Merge branch 'bugfix-2.0.x' into G33-report

thisiskeithb commented:
I don't have the permissions to link issues, but this will resolve #22886

LVD-AC commented:
This PR is about G33 delta calibration not G29 bed levelling

Guilouz commented:
Confirmed working for me.

arminth commented:
PR worked for me aswell. Anycubic Kossel Linear Plus. Todays bugfix + PR-files solved the " Autocalibration stops after first try".

Servus7 commented:
Installed Marlin on my FLSUN Q5 last weekend and Autolevelling stops after calibrating a 2cm circle in the center. This PR fixed it.
Thank You!

LVD-AC commented:
@thinkyhead

As far as I'm concerned I'm done with this but is there anything else that needs to be done in order to apply this PR to Bugfix?

- solves the confirmed bugs from the previous PR
- it calculates the max probing radius correctly now for excentric probes
- uses only already existing settings as requested
- does the calculation of `dcr` inside G33 to keep things in one place

This IMO this fixes all the known issues caused by abondoning the user defined `DELTA_CALIBRATION_RADIUS` and using a calculted `dcr` instead (transition from 1.1.9 to 2.0.0).

Remains the issue of inconsistent probe results for G33 to work properly I reported in #22853 but this is unrelated to this (it can be observed outside G33 as well) and 1st appeared in 2.0.8. I really do not have a clue of the changes that were made in 2.0.8 that may have caused this.

sjasonsmith commented:
I haven't been paying enough attention to be able to vouch for _every_ line of this change, but I verified on my printer (with a non-offset probe) that G33 does not function properly with `bugfix-2.0.x` but does work with this patch applied.

thinkyhead commented:
The confirmation helps, thanks! I'll do a final pass to see that everything is copacetic and get this merged soon.

thinkyhead committed:
Merge branch 'bugfix-2.0.x' into pr/22795

thinkyhead committed:
remove tabs, trailing spaces

thinkyhead committed:
fix unused warning

thinkyhead committed:
misc formatting
